core/lib/types/src/storage/writes/compression.rs

```diff
@@ -0,0 +1,446 @@
+        let (diff, size) = self.get_diff_and_size()?;
+
+        let mut buffer = [0u8; 32];
+        diff.to_big_endian(&mut buffer);
+
+        let diff = buffer[(32 - size)..].to_vec();
+
+        Some(diff)
+    }
+
+    fn compress_extended(&self) -> Option<Vec<u8>> {
+        match self.compress_value_only() {
+            None => None,
+            Some(compressed_val) => {
+                let mut res: Vec<u8> = vec![];
+                res.push(metadata_byte(
+                    self.output_size().unwrap(),
+                    self.operation_id(),
+                ));
+                res.extend(compressed_val);
+                Some(res)
+            }
+        }
+    }
+}
+
+struct CompressionByteNone {
+    pub new_value: U256,
+}
+
+impl CompressionByteNone {
+    fn new(new_value: U256) -> Self {
+        Self { new_value }
+    }
+}
+
+impl CompressionMode for CompressionByteNone {
+    fn operation_id(&self) -> usize {
+        0
+    }
+
+    fn output_size(&self) -> Option<usize> {
+        Some(32)
+    }
+
+    fn compress_value_only(&self) -> Option<Vec<u8>> {
+        let mut buffer = [0u8; 32];
+        self.new_value.to_big_endian(&mut buffer);
+
+        let diff = buffer[(32 - self.output_size().unwrap())..].to_vec();
+
+        Some(diff)
+    }
+
+    fn compress_extended(&self) -> Option<Vec<u8>> {
+        let mut res = [0u8; 33];
+
+        self.new_value.to_big_endian(&mut res[1..33]);
+        Some(res.to_vec())
+    }
+}
+
+fn default_passes(prev_value: U256, new_value: U256) -> Vec<Box<dyn CompressionMode>> {
+    vec![
+        Box::new(CompressionByteAdd {
+            prev_value,
+            new_value,
+        }),
+        Box::new(CompressionByteSub {
+            prev_value,
+            new_value,
+        }),
+        Box::new(CompressionByteTransform { new_value }),
+    ]
+}
+
+/// Generates the metadata byte for a given compression strategy.
+/// The metadata byte is structured as:
+/// First 5 bits: length of the compressed value
+/// Last 3 bits: operation id corresponding to the given compression used.
+fn metadata_byte(output_size: usize, operation_id: usize) -> u8 {
+    ((output_size << 3) | operation_id) as u8
+}
+
+/// For a given previous value and new value, try each compression strategy selecting the most
+/// efficient one. Using that strategy, generate the extended compression (metadata byte and compressed value).
+/// If none are found then use the full 32 byte new value with the metadata byte being `0x00`
+pub fn compress_with_best_strategy(prev_value: U256, new_value: U256) -> Vec<u8> {
+    let compressors = default_passes(prev_value, new_value);
+
+    let compressor = compressors
+        .iter()
+        .filter(|e| e.output_size().is_some())
+        .min_by(|x, y| {
+            let x_size = x.output_size();
+            let y_size = y.output_size();
+
+            x_size.cmp(&y_size)
+        });
+
+    match compressor {
+        None => CompressionByteNone::new(new_value)
```

Comment:
in theory you can get rid of this logic if you add `CompressionByteNone` to the `default` list. You'd need to do an `unwrtap` there though since you don't have information that `CompressionByteNone` never returns `None` on the type level - so ultimately up to you

Reply:
i did make an attempt to do `unwrap_or_default` with no compression as the default but it started to reach the bounds of my rust knowledge as there were complaints from the compiler about typing (`CompressionByteNone` vs `dyn CompressionMode`)